src/metabase/models/data_permissions/graph.clj

```diff
@@ -184,20 +338,98 @@
+  [model current-revision before changes]
+  (when api/*current-user-id*
+    (first (t2/insert-returning-instances! model
+                                           ;; manually specify ID here so if one was somehow inserted in the meantime in the fraction of a second since we
+                                           ;; called `check-revision-numbers` the PK constraint will fail and the transaction will abort
+                                           :id      (inc current-revision)
+                                           :before  before
+                                           :after   changes
+                                           :user_id api/*current-user-id*))))
+
+(mu/defn update-data-perms-graph!*
   "Takes an API-style perms graph and sets the permissions in the database accordingly."
-  [graph]
-  (doseq [[group-id group-changes] graph]
-    (doseq [[db-id db-changes] group-changes
-            [perm-type new-perms] db-changes]
-      (case perm-type
-        :data       (update-db-level-data-access-permissions! group-id db-id new-perms)
-        :download   (update-db-level-download-permissions! group-id db-id new-perms)
-        :data-model (update-db-level-metadata-permissions! group-id db-id new-perms)
-        :details    (update-details-perms! group-id db-id new-perms)))))
+  ([graph]
+   (doseq [[group-id group-changes] graph]
+     (doseq [[db-id db-changes] group-changes
+             [perm-type new-perms] db-changes]
+       (case perm-type
+         :data       (update-db-level-data-access-permissions! group-id db-id new-perms)
+         :download   (update-db-level-download-permissions! group-id db-id new-perms)
+         :data-model (update-db-level-metadata-permissions! group-id db-id new-perms)
+         :details    (update-details-perms! group-id db-id new-perms)))))
+
+  ;; The following arity is provided soley for convenience for tests/REPL usage
+  ([ks :- [:vector :any] new-value]
+   (update-data-perms-graph!* (assoc-in (-> api-graph :groups) ks new-value))))
+
+(mu/defn update-data-perms-graph!
+  "Takes an API-style perms graph and sets the permissions in the database accordingly. Additionally validates the revision number,
+   logs the changes, and ensures impersonations and sandboxes are consistent."
+  ([new-graph :- api.permission-graph/StrictData]
+   (let [old-graph (api-graph)
+         [old new] (data/diff (:groups old-graph) (:groups new-graph))
+         old       (or old {})
+         new       (or new {})]
+     (when (or (seq old) (seq new))
+       (log-permissions-changes old new)
+       (check-revision-numbers old-graph new-graph)
+       (check-audit-db-permissions new)
+       (t2/with-transaction [_conn]
+         (update-data-perms-graph!* new)
+         (save-perms-revision! :model/PermissionsRevision (:revision old-graph) old new)
+         (delete-impersonations-if-needed-after-permissions-change! new)
+         (delete-gtaps-if-needed-after-permissions-change! new)))))
```

Comment:
I've copied these functions over from `metabase.models.permissions` and changed that namespace to call them from here. Getting closer to removing `metabase.models.permissions` entirely.